scene/main/canvas_item.cpp

```diff
@@ -367,7 +367,12 @@ void CanvasItem::_notification(int p_what) {
 		case NOTIFICATION_WORLD_2D_CHANGED: {
 			_exit_canvas();
 			_enter_canvas();
-		}
+		} break;
+		case NOTIFICATION_PARENTED: {
+			if (!is_inside_tree()) {
+				_notify_transform();
+			}
```

Comment:
Actually, during `NOTIFICATION_PARENTED` the node is still not in the tree, tree-belonging is being set right after (and it propagates `NOTIFICATION_ENTER_TREE`):
https://github.com/godotengine/godot/blob/16a93563bfd3b02ca0a8f6df2026f3a3217f5571/scene/main/node.cpp#L1378-L1382
Meaning the `if (!is_inside_tree())` check in here is redundant / not needed. But (if I get it right) it also means that on `NOTIFICATION_ENTER_TREE`:

- This is pointless now (it should have been already invalidated at this point (by `_notify_transform()`)):

https://github.com/godotengine/godot/blob/16a93563bfd3b02ca0a8f6df2026f3a3217f5571/scene/main/canvas_item.cpp#L323

- This is still fine / needed ([relevant part should have been skipped in `_notify_transform`](https://github.com/godotengine/godot/blob/16a93563bfd3b02ca0a8f6df2026f3a3217f5571/scene/main/canvas_item.cpp#L890-L901) as the node was not yet in the tree):

https://github.com/godotengine/godot/blob/16a93563bfd3b02ca0a8f6df2026f3a3217f5571/scene/main/canvas_item.cpp#L334-L336

PS Watch out for the minotaur. :thread:

Reply:
Thanks, after digging into these parts of the code I agree, that they are not needed.